File: prototyping/OlaBuffer.m
classdef OlaBuffer < handle

    properties
        frameSize
        numOverlap

        hopSize
        delayBuffer
        addBuffer
        frameBuffers

        frameProcesser

        pDelayBuffer
        pAddBuffer
        pNewestFrame
    end

    methods
        function obj = OlaBuffer(frameSize, numOverlap, frameProcesser)
            obj.frameSize = frameSize;
            obj.numOverlap = numOverlap;
            obj.hopSize = ceil(frameSize / numOverlap);

            obj.frameProcesser = frameProcesser;

            obj.delayBuffer = zeros(frameSize, 1);
            obj.addBuffer = zeros(obj.hopSize, 1);
            obj.frameBuffers = zeros(frameSize, numOverlap);

            obj.pDelayBuffer = 1;
            obj.pAddBuffer = 1;
            obj.pNewestFrame = 1;
        end

        function block = processBlock(obj, block)

            numSamples = length(block);

            for n = 1:numSamples
        
                block(n) = obj.process(block(n));
                
            end
        end

        function x = process(obj, x)
            obj.delayBuffer(obj.pDelayBuffer) = x;
    
            if mod(obj.pDelayBuffer, obj.hopSize) == 1
    
                obj.frameBuffers(:, obj.pNewestFrame) = obj.fillNewestFrameFromDelayBuffer();

                obj.frameBuffers(:, obj.pNewestFrame) = obj.frameProcesser(obj.frameBuffers(:, obj.pNewestFrame));

                obj.fillAddBuffer();
                obj.pNewestFrame = mod(obj.pNewestFrame, obj.numOverlap) + 1;
            end
    
            x = obj.addBuffer(obj.pAddBuffer);
    
            obj.pDelayBuffer = mod(obj.pDelayBuffer, obj.frameSize) + 1;
            obj.pAddBuffer = mod(obj.pAddBuffer, obj.hopSize) + 1;
        end

        function obj = fillAddBuffer(obj)

            obj.addBuffer = obj.addBuffer * 0;
        
            frameOrder = obj.pNewestFrame:-1:(obj.pNewestFrame - obj.numOverlap + 1);
            frameOrder = mod(frameOrder - 1, obj.numOverlap) + 1;
        
            pIn = 1;
            pOut = obj.hopSize;

            for f = frameOrder
                obj.addBuffer = obj.addBuffer + obj.frameBuffers(pIn:pOut, f);
        
                pIn = pIn + obj.hopSize;
                pOut = pOut + obj.hopSize;
            end
        end

        function newestFrame = fillNewestFrameFromDelayBuffer(obj)
            
            newestFrame = zeros(obj.frameSize, 1);

            pRead = obj.pDelayBuffer;

            for n = 1:obj.frameSize
                newestFrame(n) = obj.delayBuffer(pRead);
                pRead = mod(pRead, obj.frameSize) + 1;
            end
        end
    end
    
end

File: FftBuffer/Source/OlaBuffer.cpp
//
//  OlaBuffer.cpp
//  Created by Max Henry on 2022-05-29.
//

#include "OlaBuffer.h"

OlaBuffer::OlaBuffer(int frameSize, int numOverlap)
    : frameSize(frameSize)
    , numOverlap(numOverlap)
{
    initOlaBuffer();
}

OlaBuffer::OlaBuffer()
{
    frameSize = 1024;
    numOverlap = 4;
    
    initOlaBuffer();
}

OlaBuffer::~OlaBuffer() {}

void OlaBuffer::initOlaBuffer()
{
    hopSize = frameSize / numOverlap;
    
    delayBuffer.resize(frameSize, 0.0);
    overlapAddBuffer.resize(hopSize, 0.0);
    frameBuffers.resize(numOverlap, std::vector<float>(frameSize, 0.0));
    
    pDelayBuffer = 0;
    pOverlapAddBuffer = 0;
    pNewestFrame = 0;
}

void OlaBuffer::process(float& x)
{
    delayBuffer[pDelayBuffer] = x;
    
    bool isThisHopComplete = (pDelayBuffer % hopSize == 0);
    
    if (isThisHopComplete)
    {
        std::vector<float>& newestFrame = frameBuffers[pNewestFrame];
        
        fillFrameFromDelayBuffer(newestFrame);
        
        // Virtual function to be defined by inheriting class.
        processFrameBuffers();

        // OLA calculations.
        fillOverlapAddBuffer();

        pNewestFrame = (pNewestFrame + 1) % numOverlap;
    }
    
    x = overlapAddBuffer[pOverlapAddBuffer];
    
    pDelayBuffer = (pDelayBuffer + 1) % frameSize;
    pOverlapAddBuffer = (pOverlapAddBuffer + 1) % hopSize;
}

void OlaBuffer::processFrameBuffers() {}

void OlaBuffer::fillOverlapAddBuffer()
{
    // Overlap, add and store the `hopSize`-length part where all frames overlap.
    std::fill(overlapAddBuffer.begin(), overlapAddBuffer.end(), 0.0);
    
    int pIn = 0;
    
    for (int n = 0; n < numOverlap; ++n)
    {
        // Positive-only modulus:
        //
        // https://stackoverflow.com/questions/14997165/fastest-way-to-get-a-positive-modulo-in-c-c
        int frameNo = ((pNewestFrame - n) % numOverlap + numOverlap) % numOverlap;
        
        for (int i = 0; i < hopSize; ++i)
            overlapAddBuffer[i] += frameBuffers[frameNo][pIn + i];
        
        pIn += hopSize;
    }
}

void OlaBuffer::fillFrameFromDelayBuffer(std::vector<float> &frame)
{
    int pRead = pDelayBuffer;
    
    for (int n = 0; n < frameSize; n++)
    {
        frame[n] = delayBuffer[pRead];
        pRead = (pRead + 1) % frameSize;
    }
}
